### PR TITLE
[Enhancement] Windows,System: Add AD CS audit event support

### PR DESCRIPTION
**Pull request by w0rk3r**
## Proposed commit message

```
windows, system: normalize AD CS audit events

Normalize certificate request details into winlog.adcs.* for System
Security and Windows Forwarded events. Expand native field mappings,
ECS user mappings, and event categorization so enrollment and CA
administration activity can be queried without parsing raw attributes
in search time.
```

## Summary

AD CS Security events contain certificate request and CA administration details, but they are included in `winlog.event_data.Attributes` as raw data, which lack normalization and forces us to parse them in search time, making rules more complex and expensive.

This PR normalizes selected request details into `winlog.adcs.*` in the System `security` and Windows `forwarded` data streams. It also expands native field mappings, administrative actor mappings, and event categorization.

## Issues

Part of https://github.com/elastic/detection-rules/issues/3865

## What changed

Request normalization covers Security events `4868`, `4869`, `4873`, `4874`, and `4886`–`4889`.

Normalized fields include:

* numeric request IDs
* disposition codes and recognized `CR_DISP_*` / `CERTSRV_E_*` names
* requester name and domain
* selected enrollment attributes: `CertificateTemplate`, `RequestClientInfo`, `cdc`, `rmd`, `ccm`, and `SAN`
* requested SAN values: UPN, DNS, URI, and SID URI suffixes
* extension name and OID for event `4873`

## Validation

Validated with `elastic-package` v0.126.4 and Elasticsearch 9.5.0:

* System and Windows package builds and checks
* System Security pipeline suite: 159 tests passed
* Windows Forwarded pipeline suite: 105 tests passed
* byte-identical AD CS pipelines across both packages
* captured request, disposition, extension, and administrative events, including `4892`
* existing parser edge cases
* regenerated documentation

## Checklist

* [x] I have reviewed [tips for building integrations](https://www.elastic.co/docs/extend/integrations/tips-for-building) and this pull request is aligned with them.
* [ ] I have verified that all data streams collect metrics or logs.
* [x] I have added an entry to my package's `changelog.yml` file.
* [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
* [ ] I have verified that any added dashboard complies with [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices). No dashboards are added.

## How to test this PR locally

From the repository root, test System:

```sh
cd packages/system
elastic-package build
elastic-package check
elastic-package test pipeline --data-streams security
```

From the repository root, test Windows:

```sh
cd packages/windows
elastic-package build
elastic-package check
elastic-package test pipeline --data-streams forwarded
```

The pipeline suites include the AD CS fixtures and existing events for regression coverage.

### CODE REVIEW

File: packages/windows/data_stream/forwarded/fields/winlog.yml [hunk 1]
```diff
@@ -13,6 +13,126 @@
       description: >
         A globally unique identifier that identifies the current activity. The events that are published with this identifier are part of the same activity.
 
+    - name: adcs
+      type: group
+      description: >
+        Normalized data from Active Directory Certificate Services audit events. This group is rebuilt from native fields on each normalization; existing values in this group are replaced. Native event values remain available under `winlog.event_data`.
+
```
[the remark this answers is not in the document]
Comment (w0rk3r) [reply packages/windows/data_stream/forwarded/fields/winlog.yml:20]: Changes in the Forwarded logs don't get added to the readme.

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+        }
+
+        String normalizedKey(String text, int start, int end, boolean removeSeparators) {
+          StringBuilder builder = new StringBuilder();
+          for (int index = start; index < end; index++) {
+            int character = text.charAt(index);
+            if (removeSeparators && (character == 32 || character == 9 || character == 45)) {
+              continue;
+            }
+            if (character >= 65 && character <= 90) {
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]:
```suggestion
            if ((char)'A' <= character && character <= (char)'Z') {
```

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        }
+
+        boolean asciiPrefix(String value, String prefix) {
+          if (value.length() < prefix.length()) {
+            return false;
+          }
+          for (int index = 0; index < prefix.length(); index++) {
+            int left = value.charAt(index);
+            int right = prefix.charAt(index);
+            if (left >= 65 && left <= 90) {
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]:
```suggestion
            if ((char)'A' <= left && left <= (char)'Z') {
```

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+              left += 32;
+            }
+            if (right >= 65 && right <= 90) {
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]:
```suggestion
            if ((char)'A' <= right && right <= (char)'Z') {
```

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -14,6 +14,16 @@ processors:
...
+  - pipeline:
+      name: '{{ IngestPipeline "adcs" }}'
+      if: >-
+        ctx.winlog != null &&
+        ctx.winlog.provider_name == 'Microsoft-Windows-Security-Auditing' &&
+        ctx.winlog.channel instanceof String &&
+        ctx.winlog.channel.toLowerCase() == 'security' &&
+        ctx.event != null &&
+        ctx.event.code != null &&
+        ['4868', '4869', '4873', '4874', '4886', '4887', '4888', '4889'].contains(ctx.event.code)
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/default.yml]: Was leaving 4890 out from this intentional?

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+      allow_duplicates: false
+      if: ctx._temp_adcs?.status == 'error'
+  # Preserve nonblank names; recognize dotted-decimal identifiers without decoding values.
+  - grok:
+      tag: parse_adcs_extension_name
+      field: winlog.event_data.ExtensionName
+      patterns:
+        - '\A%{ADCS_EXTENSION_NAME:_temp_adcs.request.extension.name}\z'
+      pattern_definitions:
+        ADCS_EXTENSION_NAME: '(?=[\s\S]*[^\x09-\x0D ])[\s\S]+'
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:393]: Why is there at least one `[\s\S]` required at the end? And why is it being captured? This will break parse_adcs_extension_oid which does not want a trailing whitespace.

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+          certificatetemplate: certificate_template
+          requestclientinfo: request_client_info
+          cdc: cdc
+          rmd: rmd
+          ccm: ccm
+          san: san
+        microsoft_sid_uri_prefix: 'tag:microsoft.com,2022-09-14:sid:'
+      source: |
+        boolean edgeWhitespace(int character) {
+          return character == 32 || character == 9;
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:180]: Are these the only relevant whitespace that need to be trimmed?

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml [hunk 8]
```diff
@@ -0,0 +1,451 @@
+---
+# Keep this pipeline byte-identical in:
+# - packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml
+# - packages/windows/data_stream/forwarded/elasticsearch/ingest_pipeline/security_adcs.yml
```
Comment (efd6) [inline packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Can we not use `elastic-package links` to achieve this? https://github.com/elastic/elastic-package/blob/main/docs/howto/links.md

File: packages/windows/changelog.yml [hunk 9]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "3.10.0"
+  changes:
+    - description: Normalize Active Directory Certificate Services (AD CS) audit events with native field mappings, `winlog.adcs` request fields, ECS user mappings, and event categorization.
+      type: enhancement
+      link: https://github.com/elastic/integrations/issues/21136
```
Comment (efd6) [inline packages/windows/changelog.yml]:
```suggestion
      link: https://github.com/elastic/integrations/pull/21136
```

**Review by efd6 — COMMENTED** [review]
This change would benefit from having an issue that explains the intention and links to documentation that can be used as a reference during review.

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 2 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Applied

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 3 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Applied

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 4 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Applied

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 8 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Done, switched to links

File: packages/windows/changelog.yml (hunk 9 above)
Comment (w0rk3r) [reply packages/windows/changelog.yml]: Fixed

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 6 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:393]: This part checks whether the ExtensionName field has non-ASCII-whitespace characters. The OID extraction below separately checks whether the full value matches dotted notation.

The `[\s\S]+` matches any character (whitespace + non-whitespace == every char), so it doesn't drop anything.

File: packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml (hunk 7 above)
Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:180]: It only trims spaces and tabs here, other whitespace is kept as-is. MS doesn't specify which whitespace characters to trim, and I didn't identify any other cases in the lab.

Comment (efd6) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:180]: > It only trims spaces and tabs here, other whitespace is kept as-is.

Is there a reason for this choice? Was it just that those were the only ones you saw?

Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:180]: > Was it just that those were the only ones you saw

This

Comment (efd6) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml:180]: I think it's worth extending to the full set of ASCII whitespace. The additions are relatively few.

Comment (w0rk3r) [reply packages/system/data_stream/security/elasticsearch/ingest_pipeline/adcs.yml]: Done: https://github.com/elastic/integrations/pull/21136/commits/97edf3c2c1b1ade1ff3e9b55825c2dca81d09cd1
